

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Medaka consensus training pipeline &mdash; Katuali 0.3.2 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Medaka variant-calling training pipeline" href="medaka_train_variant.html" />
    <link rel="prev" title="Custom pipelines" href="examples.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Katuali
          

          
          </a>

          
            
            
              <div class="version">
                0.3.2
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="tests.html">Basic Usage and Tests</a></li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Custom pipelines</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Medaka consensus training pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-data-specification">Input data specification</a></li>
<li class="toctree-l2"><a class="reference internal" href="#coverage-depths-specification">Coverage depths specification</a></li>
<li class="toctree-l2"><a class="reference internal" href="#creating-training-features">Creating training features</a></li>
<li class="toctree-l2"><a class="reference internal" href="#training-models">Training models</a></li>
<li class="toctree-l2"><a class="reference internal" href="#coping-with-missing-feature-files">Coping with missing feature files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="medaka_train_variant.html">Medaka variant-calling training pipeline</a></li>
<li class="toctree-l1"><a class="reference internal" href="configuration.html">Pipeline configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Frequently asked questions</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Katuali</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li>Medaka consensus training pipeline</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/medaka_train.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="medaka-consensus-training-pipeline">
<span id="medaka-train"></span><h1>Medaka consensus training pipeline<a class="headerlink" href="#medaka-consensus-training-pipeline" title="Permalink to this headline">¶</a></h1>
<p>It is possible to train and evaluate medaka consensus models starting from folders of
<code class="docutils literal notranslate"><span class="pre">.fast5</span></code> or <code class="docutils literal notranslate"><span class="pre">.fasta/q</span></code> files in a single command.</p>
<div class="section" id="input-data-specification">
<h2>Input data specification<a class="headerlink" href="#input-data-specification" title="Permalink to this headline">¶</a></h2>
<p>Read <code class="docutils literal notranslate"><span class="pre">.fast5</span></code> files should be placed under top-level folders. Multiple top-level
folders can be used, perhaps corresponding to multiple runs.</p>
<p>Within the <code class="docutils literal notranslate"><span class="pre">DATA</span></code> section of a katuali configuration file, these top-level
folders should be listed along with details of reference sequence files and sequences.
<code class="docutils literal notranslate"><span class="pre">MEDAKA_TRAIN_REGIONS</span></code> and <code class="docutils literal notranslate"><span class="pre">MEDAKA_EVAL_REGIONS</span></code> define genomic regions for training
and evaluation.</p>
<p>In the example below data from the first two top-level directories
(<code class="docutils literal notranslate"><span class="pre">MinIonRun1</span></code> and <code class="docutils literal notranslate"><span class="pre">MinIonRun2</span></code>) will be used for training using <code class="docutils literal notranslate"><span class="pre">ecoli</span></code>
and <code class="docutils literal notranslate"><span class="pre">yeast</span></code> reference sequences. Evaluation of the trained models will be
performed using the third and fourth top-level directories using the <code class="docutils literal notranslate"><span class="pre">ecoli</span></code>,
<code class="docutils literal notranslate"><span class="pre">yeast</span></code>, and <code class="docutils literal notranslate"><span class="pre">na12878_chr21</span></code> sequences.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">DATA</span><span class="p">:</span>
    <span class="s">&#39;MinIonRun1&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
    <span class="s">&#39;MinIonRun2&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
    <span class="s">&#39;GridIonRun1&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;na12878_chr21&#39;</span><span class="p p-Indicator">]</span>
    <span class="s">&#39;GridIonRun2&#39;</span><span class="p p-Indicator">:</span>
        <span class="s">&#39;REFERENCE&#39;</span><span class="p p-Indicator">:</span> <span class="s">&#39;/path/to/references.fasta&#39;</span>
        <span class="s">&#39;MEDAKA_TRAIN_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[]</span>
        <span class="s">&#39;MEDAKA_EVAL_REGIONS&#39;</span><span class="p p-Indicator">:</span> <span class="p p-Indicator">[</span><span class="s">&#39;ecoli&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;yeast&#39;</span><span class="p p-Indicator">,</span> <span class="s">&#39;na12878_chr21&#39;</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
</div>
<div class="section" id="coverage-depths-specification">
<h2>Coverage depths specification<a class="headerlink" href="#coverage-depths-specification" title="Permalink to this headline">¶</a></h2>
<p>Read depths at which to create assemblies for training are specified by the
<code class="docutils literal notranslate"><span class="pre">DEPTH</span></code> key of the katuali configuration. This list should span the range of
depths at which the model is to be used.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">DEPTHS</span><span class="p">:</span>
    <span class="p p-Indicator">[</span><span class="nv">25</span><span class="p p-Indicator">,</span> <span class="nv">50</span><span class="p p-Indicator">,</span> <span class="nv">75</span><span class="p p-Indicator">,</span> <span class="nv">100</span><span class="p p-Indicator">,</span> <span class="nv">125</span><span class="p p-Indicator">,</span> <span class="nv">150</span><span class="p p-Indicator">,</span> <span class="nv">175</span><span class="p p-Indicator">,</span> <span class="nv">200</span><span class="p p-Indicator">]</span>
</pre></div>
</div>
<p>For some datasets it my not be possible to create assemblies for all reference
sequences at all depths. To avoid katuali exiting early when such trivial failures
occur the <code class="docutils literal notranslate"><span class="pre">--keep-going</span></code> option can be used. This allows tasks to continue
unaffected by the failure of unrelated tasks.</p>
</div>
<div class="section" id="creating-training-features">
<h2>Creating training features<a class="headerlink" href="#creating-training-features" title="Permalink to this headline">¶</a></h2>
<p>To create training data (“features”) for medaka, <code class="docutils literal notranslate"><span class="pre">katuali</span></code> must:</p>
<ul class="simple">
<li><p>basecall data from all top-level directories (if <code class="docutils literal notranslate"><span class="pre">.fast5</span></code> s are provided),</p></li>
<li><p>align all basecalls to the specified reference sequences,</p></li>
<li><p>create subsampled sets of basecalls over the desired regions and depths,</p></li>
<li><p>form draft assemblies from these read sets, and finally</p></li>
<li><p>create medaka training features data and labels.</p></li>
</ul>
<p>There is a single medaka target to perform the above tasks:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali all_medaka_feat
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">Katuali</span></code> uses the <code class="docutils literal notranslate"><span class="pre">Snakemake</span></code> <code class="docutils literal notranslate"><span class="pre">--keep-going</span></code> flag instructs to continue processing tasks when
unrelated tasks fail.</p>
<p>Having run the <code class="docutils literal notranslate"><span class="pre">all_medaka_feat</span></code> target, two files will be produced for
every valid combination of dataset (top-level folder), coverage depth, and reference
sequence. For example the files:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>4bf50792/guppy/align/senterica1/25X_prop/canu_gsz_4.8m/racon/medaka_train/medaka_train.hdf
4bf50792/guppy/align/senterica1/25X_prop/canu_gsz_4.8m/racon/medaka_train/medaka_train_rc.hdf
</pre></div>
</div>
<p>will be produced for a top-level folder named <code class="docutils literal notranslate"><span class="pre">4bf5079</span></code>, a reference sequence
<code class="docutils literal notranslate"><span class="pre">senterica1</span></code> at coverage of <code class="docutils literal notranslate"><span class="pre">25</span></code>-fold.</p>
</div>
<div class="section" id="training-models">
<span id="id1"></span><h2>Training models<a class="headerlink" href="#training-models" title="Permalink to this headline">¶</a></h2>
<p>When the production of all the training data is complete, training can be commenced
by running:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>katuali all_medaka_train --keep-going
</pre></div>
</div>
<p>This step requires the the use of GPUs.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that to <code class="docutils literal notranslate"><span class="pre">tensorflow-gpu</span></code> must be installed in your medaka environment for medaka training.</p>
</div>
</div>
<div class="section" id="coping-with-missing-feature-files">
<span id="missing-feat"></span><h2>Coping with missing feature files<a class="headerlink" href="#coping-with-missing-feature-files" title="Permalink to this headline">¶</a></h2>
<p>If input datasets have insufficient coverage-depth for some of the
training regions, some training feature files will not be produced. In this
case the config flag <code class="docutils literal notranslate"><span class="pre">USE_ONLY_EXISTING_MEDAKA_FEAT</span></code> can be set to <code class="docutils literal notranslate"><span class="pre">true</span></code> to
allow katuali to train using only those features which exist already.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">USE_ONLY_EXISTING_MEDAKA_FEAT</span><span class="p">:</span> <span class="l l-Scalar l-Scalar-Plain">true</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Note that you need to first attempt to create all features with the
<code class="docutils literal notranslate"><span class="pre">medaka_train_feat</span></code> rule with <code class="docutils literal notranslate"><span class="pre">USE_ONLY_EXISTING_MEDAKA_FEAT</span></code> set to
false, and then run <code class="docutils literal notranslate"><span class="pre">all_medaka_train</span></code> with the flag set to true.</p>
</div>
<p>Refer to comments in the katuali configuration file for further details.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="medaka_train_variant.html" class="btn btn-neutral float-right" title="Medaka variant-calling training pipeline" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="examples.html" class="btn btn-neutral float-left" title="Custom pipelines" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2018-19, Oxford Nanopore Technologies

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>